backend/kernelCI_app/management/commands/process_pending_aggregations.py

```diff
@@ -0,0 +1,394 @@
+    if (
+        h_key,
+        test.test_id,
+        HardwareStatusEntityType.TEST.value,
+    ) in existing_processed:
+        return
+
+    t_pass, t_fail, t_inc = calculate_status_count(test.status)
+
+    if test.is_boot:
+        status_record["boot_pass"] += t_pass
+        status_record["boot_failed"] += t_fail
+        status_record["boot_inc"] += t_inc
+    else:
+        status_record["test_pass"] += t_pass
+        status_record["test_failed"] += t_fail
+        status_record["test_inc"] += t_inc
+
+    new_processed_entries.append(
+        ProcessedHardwareStatus(
+            hardware_key=h_key,
+            entity_id=test.test_id,
+            entity_type=HardwareStatusEntityType.TEST,
+        )
+    )
+
+
+def _process_build_status(
+    build,
+    h_key,
+    status_record,
+    existing_processed,
+    processed_builds_in_batch,
+    new_processed_entries,
+):
+    """Process build status and update status record if not already processed."""
+    if build.id.startswith(MAESTRO_DUMMY_BUILD_PREFIX):
+        return
+
+    if (
+        h_key,
+        build.id,
+        HardwareStatusEntityType.BUILD.value,
+    ) in existing_processed:
+        return
+
+    if (h_key, build.id) in processed_builds_in_batch:
+        return
+
+    b_pass, b_fail, b_inc = calculate_status_count(build.status)
+    status_record["build_pass"] += b_pass
+    status_record["build_failed"] += b_fail
+    status_record["build_inc"] += b_inc
+
+    processed_builds_in_batch.add((h_key, build.id))
+    new_processed_entries.append(
+        ProcessedHardwareStatus(
+            hardware_key=h_key,
+            entity_id=build.id,
+            entity_type=HardwareStatusEntityType.BUILD,
+        )
+    )
+
+
+def aggregate_hardware_status(
+    tests_instances: Sequence[PendingTest],
+    builds_by_id: dict[str, Builds],
+    checkouts_by_id: dict[str, Checkouts],
+):
+    hardware_status_data = {}
+    new_processed_entries = []
+
+    contexts, keys_to_check = _collect_test_contexts(
+        tests_instances, builds_by_id, checkouts_by_id
+    )
+
+    if not contexts:
+        return hardware_status_data, new_processed_entries
+
+    existing_processed = _get_existing_processed(keys_to_check)
+    processed_builds_in_batch = set()
+
+    for test, build, checkout, platform, h_key in contexts:
+        record_key = (checkout.origin, platform, checkout.id)
+
+        if record_key not in hardware_status_data:
+            hardware_status_data[record_key] = _init_status_record(
+                checkout, platform, test.compatible
+            )
+
+        status_record = hardware_status_data[record_key]
+
+        _process_test_status(
+            test, h_key, status_record, existing_processed, new_processed_entries
+        )
+
+        _process_build_status(
+            build,
+            h_key,
+            status_record,
+            existing_processed,
+            processed_builds_in_batch,
+            new_processed_entries,
+        )
```

Comment:
given that we'll never have to process a build if the test itself wasn't processed (implies the build was already done), it's better to return a boolean from `_process_test_status()`, accumulate `build` and only process it afterwards:

```py
if _process_test_status(...):
   builds_to_process.add(build)
```

where `builds_to_process` is a `set[Build]`